### PR TITLE
Removed workarounds for old compilers that don't support standard `inline`

### PR DESCRIPTION
**Pull request by seanm**

### CODE REVIEW

**Comment by seanm** [comment]
@qkoziol tells me that C99 is HDF's minimum, so hopefully therefore these changes are all OK.

File: hl/src/H5LTanalyze.c [hunk 1]
```diff
@@ -78,9 +78,11 @@
 #ifndef FLEXINT_H
 #define FLEXINT_H
 
-/* C99 systems have <inttypes.h>. Non-C99 systems may or may not. */
+/* C99 systems have <inttypes.h>. Non-C99 systems may or may not.
+ * C11 removed __STDC_LIMIT_MACROS.
+ */
 
-#if defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
+#if defined (__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L) && (__STDC_VERSION__ < 201112L)
```
Comment (derobins) [inline hl/src/H5LTanalyze.c]: This is a generated file so it should not be edited.

**Comment by gnuoyd** [comment]
On Thu, Mar 04, 2021 at 10:01:18AM -0800, Sean McBride wrote:
> @qkoziol tells me that C99 is HDF's minimum, so hopefully therefore these changes are all OK.

Thanks for this.  Hopefully CI will reveal any problems on problem
platforms like Windows.

__STDC_LIMIT_MACROS is new to me; which platform requires that?

Dave

**Comment by gnuoyd** [comment]
Oops, I see that you didn't add __STDC_LIMIT_MACROS, you just
remarked about its removal in C11. :-)  Still curious where it's
useful, if anybody knows.

Dave

File: config/cmake/H5pubconf.h.in [hunk 2]
```diff
@@ -775,25 +748,4 @@
...
-/* Define to `long' if <sys/types.h> does not define. */
-#cmakedefine H5_ptrdiff_t
-
-/* Define to `unsigned long' if <sys/types.h> does not define. */
-#cmakedefine H5_size_t
-
-/* Define to `long' if <sys/types.h> does not define. */
-#cmakedefine H5_ssize_t
-
 #endif
```
Comment (derobins) [inline config/cmake/H5pubconf.h.in]: A couple of things here:

- This file is a part of the public API, so removing symbols triggers binary compatibility problems and merges to other branches need to be done carefully.
- This file should be kept in sync with the autotools. If it's removed in CMake, it needs to be removed in configure.ac as well.
- HDF5 is supported on systems with dodgy C implementations so we tend to be conservative about ripping out old work-arounds and hacks.

I'm definitely excited to remove old cruft, but we do need to be careful when doing so. At the very least, we need to make sure that CMake and the Autotools stay in sync.

File: hl/src/H5LTanalyze.c (hunk 1 above)
Comment (seanm) [reply hl/src/H5LTanalyze.c]: Ah yes, will remove.  I wonder if newer flex fixes this...

File: config/cmake/H5pubconf.h.in (hunk 2 above)
Comment (seanm) [reply config/cmake/H5pubconf.h.in]: I grepped the entire codebase for `H5_ssize_t` and this is the only occurrence anywhere. That's why I removed it, figured it was leftover from a previous purge.

**Comment by derobins** [comment]
C99 is basically the agreed-upon minimum for HDF5, but we still support older versions of Visual Studio with dodgy C99 support, so we just can't rip out all pre-C99 hacks. I don't know if it's still the case, but h5py on Windows was built with an ancient version of Visual Studio and alienating that community probably isn't wise. In general, though, a lot of the hacks can go.

I think  __STDC_LIMIT_MACROS was a compatibility macro for use with pre-C++11 code. It should no longer be needed and, in fact, causes symbol redefinition warnings on VS2019.

File: config/cmake/H5pubconf.h.in (hunk 2 above)
Comment (derobins) [reply config/cmake/H5pubconf.h.in]: Removing cruft from old operating systems and compilers is a never-ending task :)

**Comment by derobins** [comment]
> On Thu, Mar 04, 2021 at 10:01:18AM -0800, Sean McBride wrote: @qkoziol tells me that C99 is HDF's minimum, so hopefully therefore these changes are all OK.
> Thanks for this. Hopefully CI will reveal any problems on problem platforms like Windows.
> […](#)
> __STDC_LIMIT_MACROS is new to me; which platform requires that? Dave

The CI on github will probably not pick up C99 issues with elderly VS. We only test VS2019 there. Allen tests a much wider variety of VS flavors, but I think that only happens once things land in one of the maintenance branches (develop, hdf5_1_12, etc.).

**Comment by seanm** [comment]
What's the minimum Visual Studio supported?

File: config/cmake/H5pubconf.h.in (hunk 2 above)
Comment (seanm) [reply config/cmake/H5pubconf.h.in]: So is this bit ok?

Comment (derobins) [reply config/cmake/H5pubconf.h.in]: I would probably split this into two PRs: One PR with the internal inline code changes. Another with the CMake changes.

The internal inline code PR can be propagated (more) easily to other branches without worrying about binary compatibility issues.

The CMake PR will probably only be able to go to develop (1.13+) as we don't remove public symbols in maintenance branches. We do need to keep the Autotools and CMake in sync, but if you create a PR with the CMake changes, I'll create a PR with the Autotools changes that you can merge into your PR for Larry to do a squash merge.

**Comment by derobins** [comment]
Our Windows testing person says that VS2015 is the current minimum (which is higher than I thought - I thought we were still stuck on 2012).

File: config/cmake/H5pubconf.h.in (hunk 2 above)
Comment (derobins) [reply config/cmake/H5pubconf.h.in]: I feel like we should also remove the CMake code that checks for these things, to fully expunge the cruft.

Comment (derobins) [reply config/cmake/H5pubconf.h.in]: Example: The inline checks in config/cmake_ext_mod/HDFTests.c

Comment (seanm) [reply config/cmake/H5pubconf.h.in]: ok split to: https://github.com/HDFGroup/hdf5/pull/431 let's discuss there

**Comment by seanm** [comment]
OK, this contains only the `inline` cleanup commit now.

File: config/cmake/H5pubconf.h.in [hunk 3]
```diff
@@ -480,12 +477,6 @@
 /* Define to 1 if you have the `_scrsize' function. */
 #cmakedefine H5_HAVE__SCRSIZE @H5_HAVE__SCRSIZE@
 
-/* Define if the compiler understands __inline */
-#cmakedefine H5_HAVE___INLINE @H5_HAVE___INLINE@
-
-/* Define if the compiler understands __inline__ */
-#cmakedefine H5_HAVE___INLINE__ @H5_HAVE___INLINE__@
-
 /* Define if the library will ignore file locks when disabled */
```
Comment (derobins) [inline config/cmake/H5pubconf.h.in]: I would leave these symbols here and pull them in the CMake PR. Having them here makes it difficult to propagate the inline code changes.

File: src/H5private.h [hunk 4]
```diff
@@ -190,23 +190,6 @@
...
-#elif defined(H5_HAVE_INLINE)
-/* GNU, C++
- * Use "inline" as a last resort on the off-chance that there will
- * be C++ problems.
- */
-#define H5_INLINE inline
-#else
-#define H5_INLINE
-#endif /* inline choices */
-
```
Comment (derobins) [inline src/H5private.h:209]: Since this is a private file, these deletions can stay.

**Review by qkoziol — APPROVED** [review]

**Review by derobins — CHANGES_REQUESTED** [review]
The stuff in H5pubconf needs to be removed to make this easier to merge to downstream maintenance branches.

File: config/cmake/H5pubconf.h.in (hunk 3 above)
Comment (seanm) [reply config/cmake/H5pubconf.h.in]: I'm a bit confused by this.  You want me to deliberately make a PR (this one) that only half-removes the old inline workarounds, and leaves that `#cmakedefine` statement as dead code?

Comment (derobins) [reply config/cmake/H5pubconf.h.in]: Yes. Exactly that. We can change the code but leave the configure/CMake bits in older branches.

Comment (seanm) [reply config/cmake/H5pubconf.h.in]: OK done.

**Review by derobins — APPROVED** [review]

**Pull request merged**